### DISCUSSION TITLE
How about using this tool as PaaS offering

### DISCUSSION THREAD

**Discussion by amitjattan** [Azure/ipam#67] [Ideas]
@msftgits @DCMattyG  Looks like currently tool is deployed as owned solution by customer and it's hard to accommodate new features in the tool when added in this repository, how do you suggests consumers to update tool once deployed in their environments?

**Comment by DCMattyG** [comment]
Hi @amitjattan, this is a great question!

If a customer deploys the standard offering, the containers are hosted in a public Azure Container Registry. As the containers are updated, the only thing users have to do is simply restart the App Service, and the new containers will be pulled down. In the near future, I'll probably add in some automation around this restart so it "just happens".

If a customer chooses to deploy this solution in a Private ACR, then they will need to pull down the updated code from GitHub and build the new containers into their ACR's on a regular basis.

Did that help to answer your question completely?

**Reply by amitjattan** [reply]
Thanks @DCMattyG yeah this is helpful... One more question, seems like the reservations and other data like blocks, address spaces are currently stored Cosmos DB. How to handle the situation if there are upgrades in tool associated with DB upgrades. assuming in such cases we will have to upgrade DB also which would have some table, columns updates. Wondering if there would be scripts available for such scenarios. 

FYI...We are currently planning to give this solution a dry run and see what it offers.

**Reply by DCMattyG** [reply]
Good morning @amitjattan!

To indirectly answer your question, we are indeed using CosmosDB, which is a NoSQL solution, meaning there really isn't a "Schema" as defined by the database, but instead any schema-like format would be enforced at the application layer. At the end of the day, it's just a bunch of JSON documents. That combined with the fact that we use a tiny, dedicated CosmosDB for IPAM, are reasons you won't have to worry about a situation as you've outlined above.

To directly answer your question, if there are any database level changes, we are currently handling that during the Engine startup. The Engine will check the DB to see if anything needs to be "upgraded", make the changes, and then continue. There shouldn't ever be any reason to externally upgrade the database, or make any changes.

Hoped that help to clarify things!

**Reply by amitjattan** [reply]
That's great news thanks @DCMattyG !! So upgrades should be done normal ways ( Container version updated in ACR and restart the App service) and no impacts should happen on existing data stored on deployed IPAM service?

**Reply by DCMattyG** [reply]
That is 100% correct!

**Reply by amitjattan** [reply]
Hi @DCMattyG I have deployed the app in my environment. But seems like it not making progress when running on App Service. Always getting this error. Also, there are not logs being generated on console or in log streams. Not sure what is missed.

:( Application Error
If you are the application administrator, you can access the [diagnostic resources](https://ipam-as.scm.azurewebsites.net/detectors).

**Reply by DCMattyG** [reply]
Hi @amitjattan, I won't have access to any of your logs from your Azure environment.

When you deployed, a logs folder was created that has all of the deployment logs. You can email them to me at Matthew.Garrett@microsoft.com for review and I'm happy to help you further. Please make sure that you've met the prerequisites prior to deploying as well:

https://azure.github.io/ipam/#/deployment/README?id=prerequisites

**Reply by amitjattan** [reply]
Yeah I have all the required access needed for me, So I have deployed the app in two parts 

**AppsOnly mode - Here are the logs** 

**********************
PowerShell transcript start
Start time: 20230112165316
Configuration Name: 
Host Application: /usr/local/microsoft/powershell/7/pwsh.dll
Process ID: 79063
PSVersion: 7.3.1
PSEdition: Core
GitCommitId: 7.3.1
OS: Darwin 21.6.0 Darwin Kernel Version 21.6.0: Mon Aug 22 20:17:10 PDT 2022; root:xnu-8020.140.49~2/RELEASE_X86_64
Platform: Unix
PSCompatibleVersions: 1.0, 2.0, 3.0, 4.0, 5.0, 5.1.10032.0, 6.0.0, 6.1.0, 6.2.0, 7.0.0, 7.1.0, 7.2.0, 7.3.1
PSRemotingProtocolVersion: 2.3
SerializationVersion: 1.1.0.1
WSManStackVersion: 3.0
**********************

NOTE: IPAM Deployment Type: AppsOnly
INFO: Fetching Tenant ID from Azure PowerShell SDK
INFO: Fetching Azure Cloud type from Azure PowerShell SDK
INFO: Creating Azure IPAM UI Application
INFO: Creating Azure IPAM Engine Application
INFO: Printing Azure IPAM Engine API Endpoint b15542ea-d1a4-4f0c-af4b-23d7346b9daf
INFO: Updating Azure IPAM Engine API Endpoint
INFO: Updating Azure IPAM UI Application Resource Access
INFO: Creating Azure IPAM UI Service Principal
INFO: Creating Azure IPAM Engine Service Principal

New-AzRoleAssignment: 
./local/share/powershell/Modules/Az.Resources/6.5.1/MSGraph.Autorest/custom/New-AzADServicePrincipal.ps1:752
Line |
 752 |        $ra = New-AzRoleAssignment @param
     |              ~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Operation returned an invalid status code 'Forbidden'

New-AzRoleAssignment: 
/.local/share/powershell/Modules/Az.Resources/6.5.1/MSGraph.Autorest/custom/New-AzADServicePrincipal.ps1:752
Line |
 752 |        $ra = New-AzRoleAssignment @param
     |              ~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Operation returned an invalid status code 'Forbidden'

INFO: Creating Azure IPAM Engine Secret
INFO: Azure IPAM Engine & UI Applications/Service Principals created successfully
INFO: Logging in to Microsoft Graph
INFO: Granting admin consent for Microsoft Graph API permissions assigned to IPAM UI application
INFO: Admin consent for Microsoft Graph API permissions granted successfully
INFO: Granting admin consent to the IPAM UI application for exposed API from the IPAM Engine application
INFO: Admin consent for IPAM Engine exposed API granted successfully
INFO: Granting admin consent for Azure Service Management API permissions assigned to IPAM Engine application
INFO: Admin consent for Azure Service Management API permissions granted successfully
INFO: Populating Bicep parameter file for infrastructure deployment
INFO: Bicep parameter file populated successfully
INFO: Azure IPAM Solution deployed successfully

**********************
PowerShell transcript end
End time: 20230112165342
**********************

**Bicep Deployment**

**********************
PowerShell transcript start
Start time: 20230112165638

Configuration Name: 

Host Application: /usr/local/microsoft/powershell/7/pwsh.dll
Process ID: 79063
PSVersion: 7.3.1
PSEdition: Core
GitCommitId: 7.3.1
OS: Darwin 21.6.0 Darwin Kernel Version 21.6.0: Mon Aug 22 20:17:10 PDT 2022; root:xnu-8020.140.49~2/RELEASE_X86_64
Platform: Unix
PSCompatibleVersions: 1.0, 2.0, 3.0, 4.0, 5.0, 5.1.10032.0, 6.0.0, 6.1.0, 6.2.0, 7.0.0, 7.1.0, 7.2.0, 7.3.1
PSRemotingProtocolVersion: 2.3
SerializationVersion: 1.1.0.1
WSManStackVersion: 3.0
**********************

NOTE: IPAM Deployment Type: TemplateOnly
INFO: Validating Azure Region selected for deployment
INFO: Azure Region validated successfully
INFO: Importing values from Bicep parameters file
INFO: Successfully import Bicep parameter values for Full deployment
INFO: Deploying IPAM bicep templates
WARNING: /Users/s004700/work/ipam/deploy/managedIdentity.bicep(2,7) : Warning no-unused-params: Parameter "contributorAssignmentName" is declared but never used. [https://aka.ms/bicep/linter/no-unused-params]
/Users/s004700/work/ipam/deploy/managedIdentity.bicep(11,7) : Warning no-unused-params: Parameter "managedIdentityOperatorAssignmentName" is declared but never used. [https://aka.ms/bicep/linter/no-unused-params]
/Users/s004700/work/ipam/deploy/managedIdentity.bicep(14,5) : Warning no-unused-vars: Variable "contributorId" is declared but never used. [https://aka.ms/bicep/linter/no-unused-vars]
/Users/s004700/work/ipam/deploy/managedIdentity.bicep(16,5) : Warning no-unused-vars: Variable "managedIdentityOperatorId" is declared but never used. [https://aka.ms/bicep/linter/no-unused-vars]
/Users/s004700/work/ipam/deploy/main.bicep(5,7) : Warning no-unused-params: Parameter "guid" is declared but never used. [https://aka.ms/bicep/linter/no-unused-params]
INFO: IPAM bicep templates deployed successfully
INFO: Azure IPAM Solution deployed successfully

POST DEPLOYMENT TASKS:
##############################################
Navigate In Browser To:
https://portal.azure.com/#view/Microsoft_AAD_RegisteredApps/ApplicationMenuBlade/~/Authentication/appId/XXXX
Change 'Redirect URI' To:
https://XXXX.azurewebsites.net
##############################################

**********************
PowerShell transcript end
End time: 20230112165923
**********************

**Reply by DCMattyG** [reply]
It appears you do note have the ```Microsoft.Authorization/roleAssignments/write``` of the Root Management Group, hence the automation is unable to assign the roles needed for the IPAM Engine to function. You can see from the logs that **New-AzRoleAssignment** fails.

**Reply by DCMattyG** [reply]
BTW, if you're going to run the entire deployment yourself, you don't need to do an *Apps Only* deployment followed by a *Infra Only* deployment. You can just do a regular *Full* deployment and this will perform both operations at one time:

https://azure.github.io/ipam/#/deployment/README?id=full-deployment

**Reply by amitjattan** [reply]
Hi @DCMattyG thanks for pointing the error, I have manually assigned the reader access on the root. And just tried to run the app in local first. It seems to be working for me. However, I am not able to see the Admin Screen. Not sure why

![image](https://user-images.githubusercontent.com/26720834/212377665-192590f1-2222-4da7-86fe-69ae570c8949.png)

**Reply by DCMattyG** [reply]
Hi @amitjattan, it appears you are navigating to "localhost"....am I reading that correctly?

**Reply by amitjattan** [reply]
Yes this is correct @DCMattyG

**Reply by amitjattan** [reply]
And this is the error it shows in logs, if I just run with context localhost/admin/admins and try to save it

"PUT /api/admin/admins HTTP/1.0" 403 Forbidden

**Reply by DCMattyG** [reply]
So how are you running this? It's designed to be run in Azure App Services (assuming you are using the provided deploy script). I would assume your URL would be *https://&lt;something&gt;.azurewebsites.net*. Can you help me better how exactly this is deployed in your circumstance?

**Reply by amitjattan** [reply]
So I have previousely executed the the ./deploy.ps1 --AppsOnly script and I was able to manage All SPNs/Apps created with correct permissions and then I have updated the ipam-ui-app redirect url to http://localhost

As a second step, where I have deployed the bicep templates, which created cosmos DB and key vault for me, I had just used those instances as it is for my local docker-compose and then updated the docker-compose.prod.yml with the secrets directly and ran that in local.
